raiden/utils/migrations/v18_to_v19.py

```diff
@@ -0,0 +1,218 @@
+            query_records,
+        )
+        for entry in pool_generator:
+            updated_state_changes.append(entry)
+
+        # Finally update the DB with a batched executemany()
+        storage.update_state_changes(updated_state_changes)
+
+
+def _add_blockhash_to_events(storage: SQLiteStorage, cache: BlockHashCache) -> None:
+    """Adds blockhash to all ContractSendXXX events"""
+    batch_query = storage.batch_query_event_records(
+        batch_size=500,
+        filters=[('_type', 'raiden.transfer.events.ContractSend%')],
+    )
+    for events_batch in batch_query:
+        updated_events = []
+        for event in events_batch:
+            data = json.loads(event.data)
+            assert 'triggered_by_block_hash' not in data, 'v18 events cant contain blockhash'
+            # Get the state_change that triggered the event and if it has
+            # a block number get its hash. If not fall back to latest.
+            matched_state_changes = storage.get_statechanges_by_identifier(
+                from_identifier=event.state_change_identifier,
+                to_identifier=event.state_change_identifier,
+            )
+            result_length = len(matched_state_changes)
+            msg = 'multiple state changes should not exist for the same identifier'
+            assert result_length == 1, msg
+
+            statechange_data = json.loads(matched_state_changes[0])
+            if 'block_hash' in statechange_data:
+                block_hash = statechange_data['block_hash']
+            elif 'block_number' in statechange_data:
+                block_number = int(statechange_data['block_number'])
+                block_hash = cache.get(block_number)
+            data['triggered_by_block_hash'] = block_hash
+            updated_events.append((
+                json.dumps(data),
+                event.event_identifier,
+            ))
+
+        storage.update_events(updated_events)
+
+
+def _transform_snapshot(
+        raw_snapshot: Dict[Any, Any],
+        storage: SQLiteStorage,
+        cache: BlockHashCache,
+) -> str:
+    """Upgrades a single snapshot by adding the blockhash to it and to any pending transactions"""
+    snapshot = json.loads(raw_snapshot)
+    block_number = int(snapshot['block_number'])
+    snapshot['block_hash'] = cache.get(block_number)
+
+    pending_transactions = snapshot['pending_transactions']
+    new_pending_transactions = []
+    for transaction_data in pending_transactions:
+        if 'raiden.transfer.events.ContractSend' not in transaction_data['_type']:
+            raise InvalidDBData(
+                "Error during v18 -> v19 upgrade. Chain state's pending transactions "
+                "should only contain ContractSend transactions",
+            )
+
+        # For each pending transaction find the corresponding DB event record.
+        event_record = storage.get_latest_event_by_data_field(
+            filters=transaction_data,
```

Comment:
Hummm, I don't think this work as expected, `transaction_data` will be a `ContractSend` event, added to the queue here:

https://github.com/raiden-network/raiden/blob/ceff49d7d7dec9b2ded273aacbfd7882a7fd73e3/raiden/transfer/node.py#L1205-L1206

Some of these events contain nested data, e.g. `balance_proof` for the `ContractSendChannelClose`:

https://github.com/raiden-network/raiden/blob/ceff49d7d7dec9b2ded273aacbfd7882a7fd73e3/raiden/transfer/events.py#L84-L92

This means that the value for that key will be a dictionary, and when the `filters` dictionary is processed `value` will actually be the `repr` for the dictionary:

https://github.com/raiden-network/raiden/blob/ceff49d7d7dec9b2ded273aacbfd7882a7fd73e3/raiden/storage/sqlite.py#L219-L222

Which may not be valid JSON (and even if it was, I'm not sure the query would work).

IMO we should:

- Create a proper type for the filter `Dict[str, str]`, anything else is broken
- Add an assert, to verify the above type is used
- Fix this to properly encode the `ContractSend` event as a filter

In addition to the above, we have another problem, the even in the snapshot *may not be equal the event in the database*, this is true because we are doing updates. In this particular case, everything will work, but I can see people in the future just copying and pasting this code, and trying to be tidy and do the equivalent of `del statechange_data['block_hash']` in the function `_add_blockhash_to_events`, this would break the code, IMO this should be more explicit to avoid the above.

Reply:
Example code to filter events with nested data:

```python
def filter_from_dict(d):
    filter_ = dict()

    for k, v in current.iteritems():
        if isinstance(v, dict):
            for sub, v2 in flatten(v).iteritems():
                filter_[f'{k}.{sub}'] = v2
        else:
            filter_[k] = v

    return filter_
```